server/src/main/java/org/elasticsearch/cluster/metadata/LifecycleExecutionState.java

```diff
@@ -287,15 +287,16 @@ public Map<String, String> asMap() {
         return Collections.unmodifiableMap(result);
     }
 
-    public static String truncateWithExplanation(String input) {
-        if (input != null && input.length() > MAXIMUM_STEP_INFO_STRING_LENGTH) {
-            return Strings.cleanTruncate(input, MAXIMUM_STEP_INFO_STRING_LENGTH)
-                + "... ("
-                + (input.length() - MAXIMUM_STEP_INFO_STRING_LENGTH)
-                + " chars truncated)";
-        } else {
-            return input;
+    public static String potentiallyTruncateLongJsonWithExplanation(String json) {
+        if (json == null) {
+            return null;
         }
+        // Avoid no-op truncations: we must append `"}` (2 chars) to the end to keep JSON valid
+        if (json.length() <= MAXIMUM_STEP_INFO_STRING_LENGTH + 2) {
+            return json;
+        }
+        final int actuallyRemovedCharacters = json.length() - MAXIMUM_STEP_INFO_STRING_LENGTH - 2;
+        return Strings.cleanTruncate(json, MAXIMUM_STEP_INFO_STRING_LENGTH) + "... (" + actuallyRemovedCharacters + " chars truncated)\"}";
```

Comment:
Do we actually care about the 2 extra chars? This truncation is just a way to avoid very long strings/json; it doesn't actually care about being exactly `MAXIMUM_STEP_INFO_STRING_LENGTH`. Couldn't we just leave the `2` out and slightly simplify this method?